packages/extension-vscode/src/server.ts

```diff
@@ -28,6 +55,12 @@ connection.onDidChangeWatchedFiles(async () => {
     }));
 });
 
+connection.onCodeAction(quickFixActionProvider.provideCodeActions.bind(quickFixActionProvider));
```

Comment:
After reading through this, I think we can move the CodeActionProvider to the extension.ts side which would allow us to reduce message passing and modifications to the language server (since we can also read the current list of diagnostics from that side).

See this vscode sample for an example of registering a CodeActionProvider from the extension host:
https://github.com/microsoft/vscode-extension-samples/blob/30c1740717acdc7e4aabc783d07bd83e21857320/code-actions-sample/src/extension.ts#L22

Reply:
Revisiting since I've been looking at workspace edits. Apparently CodeActions from the language server can include an `edit` instead of a `command`. This could potentially allow us to specify the changes we want VS Code to apply to a file from the language server without having to add custom handling in the extension host.

I think either this or my above suggestion would help reduce the contract between the host and the language server.